src/debug/debug_stream/debug_stream_thread_info.c

```diff
@@ -0,0 +1,387 @@
+#endif
+
+static void thread_info_get(int core, struct record_buf *rec)
+{
+	k_thread_runtime_stats_t core_stats;
+	struct user_data ud = {
+		.core = core,
+		.rec = rec,
+		.thread_count = 0,
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+		.active_threads = { NULL },
+		.previous = &previous[core],
+#endif
+	};
+	struct thread_info_record_hdr *hdr;
+	uint8_t load = 0;
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+	int ret = k_thread_runtime_stats_cpu_get(core, &core_stats);
+
+	if (ret == 0) {
+		uint32_t active_cycles = (uint32_t) (core_stats.total_cycles -
+						     ud.previous->active);
+		uint32_t all_cycles = (uint32_t) (core_stats.execution_cycles -
+						  ud.previous->all);
+
+		ud.stats_valid = true;
+		load = (uint8_t) ((255LLU * active_cycles) / all_cycles);
+		LOG_DBG("Core %u load %u / %u total %llu / %llu", core,
+			active_cycles, all_cycles,
+			core_stats.total_cycles, core_stats.execution_cycles);
+		ud.previous->active = core_stats.total_cycles;
+		ud.previous->all = core_stats.execution_cycles;
+		ud.all_cycles = all_cycles;
+	}
+#endif
+
+	hdr = (struct thread_info_record_hdr *) ud.rec->buf;
+	rec->w_ptr = sizeof(*hdr);
+	hdr->hdr.id = DEBUG_STREAM_RECORD_ID_THREAD_INFO;
+	hdr->load = load;
+	/* This is best effort debug tool. Unlocked version should be fine. */
+	k_thread_foreach_unlocked_filter_by_cpu(core, thread_info_cb, &ud);
+
+	cleanup_old_thread_cycles(&ud);
+
+	hdr->thread_count = ud.thread_count;
+	hdr->hdr.size = SOF_DIV_ROUND_UP(rec->w_ptr, sizeof(hdr->hdr.data[0]));
+	debug_stream_slot_send_record(&hdr->hdr);
+}
+
+static void thread_info_run(void *cnum, void *a, void *b)
+{
+	int core = (int) cnum;
+	struct record_buf rec = {
+		.size = THREAD_INFO_INITIAL_RECORD_BUFFER_SIZE,
+		.w_ptr = 0,
+	};
+
+	rec.buf = rmalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, rec.size);
+	if (!rec.buf) {
+		LOG_ERR("malloc failed");
+		return;
+	}
+
+	for (;;) {
+		thread_info_get(core, &rec);
+		k_sleep(K_SECONDS(CONFIG_SOF_DEBUG_STREAM_THREAD_INFO_INTERVAL));
+	}
+}
+
+#define THREAD_STACK_SIZE (2048)
+static K_THREAD_STACK_ARRAY_DEFINE(info_thread_stacks, CONFIG_MP_MAX_NUM_CPUS,
+				   THREAD_STACK_SIZE);
+static struct k_thread info_thread[CONFIG_MP_MAX_NUM_CPUS];
+
+static int thread_info_start(void)
+{
+	uint32_t i;
+
+	for (i = 0; i < ARRAY_SIZE(info_thread); i++) {
+		char name[24];
+		k_tid_t tid = NULL;
+		int ret;
+
+		tid = k_thread_create(&info_thread[i], info_thread_stacks[i],
+				      THREAD_STACK_SIZE, thread_info_run,
+				      (void *) i, NULL, NULL,
+				      K_LOWEST_APPLICATION_THREAD_PRIO, 0,
+				      K_FOREVER);
+		if (!tid) {
+			LOG_ERR("k_thread_create() failed for core %u", i);
+			continue;
+		}
+		ret = k_thread_cpu_pin(tid, i);
+		if (ret < 0) {
+			LOG_ERR("Pinning thread to code core %u", i);
+			k_thread_abort(tid);
+			continue;
+		}
+		snprintf(name, sizeof(name), "%u thread info", i);
+		ret = k_thread_name_set(tid, name);
+		if (ret < 0)
+			LOG_INF("k_thread_name_set failed: %d for %u", ret, i);
```

Comment:
There is only two reasons why k_thread_name_set() would fail. Either the  thread is not found (should not happen here), or CONFIG_THREAD_NAME=n. So this is mostly just a gentle indication that the thread names are configured.

Reply:
Ack